File: src/commands/account.ts
import z from "zod";
import { Account } from "@/lib/types";
import { newAccountSchema } from "@/lib/schemas";
import { invokeTauri, logger } from "@/adapters";

type NewAccount = z.infer<typeof newAccountSchema>;

export const getAccounts = async (): Promise<Account[]> => {
  try {
    return invokeTauri("get_accounts");
  } catch (error) {
    logger.error("Error fetching accounts.");
    throw error;
  }
};

// createAccount
export const createAccount = async (account: NewAccount): Promise<Account> => {
  try {
    return invokeTauri("create_account", { account: account });
  } catch (error) {
    logger.error("Error creating account.");
    throw error;
  }
};

// updateAccount
export const updateAccount = async (account: NewAccount): Promise<Account> => {
  try {
    const { currency: _currency, ...updatedAccountData } = account;
    return invokeTauri("update_account", { accountUpdate: updatedAccountData });
  } catch (error) {
    logger.error("Error updating account.");
    throw error;
  }
};

// deleteAccount
export const deleteAccount = async (accountId: string): Promise<void> => {
  try {
    await invokeTauri("delete_account", { accountId });
    return;
  } catch (error) {
    logger.error("Error deleting account.");
    throw error;
  }
};

File: src-tauri/src/commands/account.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use log::{debug, error, info};
use tauri::State;

use wealthfolio_core::accounts::{Account, AccountUpdate, NewAccount, TrackingMode};

#[tauri::command]
pub async fn get_accounts(state: State<'_, Arc<ServiceContext>>) -> Result<Vec<Account>, String> {
    debug!("Fetching active accounts...");
    state
        .account_service()
        .get_all_accounts()
        .map_err(|e| format!("Failed to load accounts: {}", e))
}

#[tauri::command]
pub async fn create_account(
    account: NewAccount,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Account, String> {
    debug!("Adding new account...");
    // Domain events handle recalculation automatically
    state
        .account_service()
        .create_account(account)
        .await
        .map_err(|e| {
            error!("Failed to add new account: {}", e);
            format!("Failed to add new account: {}", e)
        })
}

#[tauri::command]
pub async fn update_account(
    account_update: AccountUpdate,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Account, String> {
    debug!("Updating account {:?}...", account_update.id);

    // Domain events handle recalculation automatically
    state
        .account_service()
        .update_account(account_update.clone())
        .await
        .map_err(|e| format!("Failed to update account {:?}: {}", account_update.id, e))
}

#[tauri::command]
pub async fn delete_account(
    account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!("Deleting account {}...", account_id);
    // Domain events handle recalculation automatically
    state
        .account_service()
        .delete_account(&account_id)
        .await
        .map_err(|e| {
            error!("Failed to delete account {}: {}", account_id, e);
            e.to_string()
        })
}

/// Switches an account's tracking mode with proper handling of snapshot sources.
/// Updates account meta with the new tracking mode.
#[tauri::command]
pub async fn switch_tracking_mode(
    account_id: String,
    new_mode: TrackingMode,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<(), String> {
    debug!(
        "Switching tracking mode for account {} to {:?}",
        account_id, new_mode
    );

    // Get the current account for meta update
    let account = state
        .account_service()
        .get_account(&account_id)
        .map_err(|e| format!("Failed to get account: {}", e))?;

    // Update the account meta with the new tracking mode
    let new_meta = wealthfolio_core::accounts::set_tracking_mode(account.meta.clone(), new_mode);

    let account_update = AccountUpdate {
        id: Some(account_id.clone()),
        name: account.name,
        account_type: account.account_type,
        group: account.group,
        is_default: account.is_default,
        is_active: account.is_active,
        platform_id: account.platform_id,
        account_number: account.account_number,
        meta: Some(new_meta),
        provider: account.provider,
        provider_account_id: account.provider_account_id,
    };

    // Domain events handle recalculation and broker sync automatically
    state
        .account_service()
        .update_account(account_update)
        .await
        .map_err(|e| format!("Failed to update account: {}", e))?;

    info!(
        "Successfully switched tracking mode for account {} to {:?}",
        account_id, new_mode
    );

    Ok(())
}
